File: src/python/day12.py
example_input = """
0:
###
##.
##.

1:
###
##.
.##

2:
.##
###
##.

3:
##.
###
##.

4:
###
#..
###

5:
###
.#.
###

4x4: 0 0 0 0 2 0
12x5: 1 0 1 0 2 2
12x5: 1 0 1 0 3 2
"""

class Present:
  def __init__(self, shape):
    self.shape = shape

  def __repr__(self):
    return "\n" + "\n".join(["".join(row) for row in self.shape]) + "\n"

  def width(self):
    return max([len(row) for row in self.shape])

  def height(self):
    return len(self.shape)

  def cells(self) -> int:
    return sum([1 for row in self.shape for cell in row if cell == '#'])


class Region:
  def __init__(self, width, height, presents):
    self.width = width
    self.height = height
    self.presents = presents

  def __repr__(self):
    return f"{self.width}x{self.height}: {self.presents}"


def parse_input(input):
  presents = input.split("\n\n")
  regions = presents.pop()
  present_list = []
  region_list = []

  for present in presents:
    lines = present.split("\n")
    shape = [list(line) for i, line in enumerate(lines) if i != 0]
    present_list.append(Present(shape))

  for region in regions.split("\n"):
    dimensions, presents = region.split(": ")
    width, height = dimensions.split("x")
    presents_region = [int(p) for p in presents.split(" ")]
    region_list.append(Region(int(width), int(height), presents_region))

  return present_list, region_list
    

def region_fits(region, present_list):
  return region.width * region.height >= sum([p * q.cells() for p, q in zip(region.presents, present_list)])


def part1(input):
  presents, regions = input
  how_many_fit = 0
  for region in regions:
    if region_fits(region, presents):
      how_many_fit += 1
  return how_many_fit
    
def part2(input):
  return "\u2605"
    

input = open("input/day12.txt", "r").read()
# input = example_input.strip()
print("PART 1 =", part1(parse_input(input)))
print("PART 2 =", part2(parse_input(input)))

File: src/rust/day12.rs
struct Present {
    shape: Vec<Vec<char>>,
}

impl Present {
    fn cells(&self) -> usize {
        self.shape
            .iter()
            .flat_map(|row| row.iter())
            .filter(|&&cell| cell == '#')
            .count()
    }
}

#[derive(Debug)]
struct Region {
    width: usize,
    height: usize,
    presents: Vec<usize>,
}

fn parse_input(input: &str) -> (Vec<Present>, Vec<Region>) {
    let mut sections: Vec<&str> = input.trim().split("\n\n").collect();
    let regions_str = sections.pop().unwrap();

    // Parse Presents
    let presents = sections
        .iter()
        .map(|section| {
            let shape = section.lines().skip(1).map(|line| line.chars().collect()).collect();
            Present { shape }
        })
        .collect();

    // Parse Regions
    let regions = regions_str
        .lines()
        .map(|line| {
            let (dimensions, counts) = line.split_once(": ").unwrap();
            let (w, h) = dimensions.split_once('x').unwrap();
            let presents_region = counts.split_whitespace().map(|n| n.parse().unwrap()).collect();

            Region {
                width: w.parse().unwrap(),
                height: h.parse().unwrap(),
                presents: presents_region,
            }
        })
        .collect();

    (presents, regions)
}

fn region_fits(region: &Region, present_list: &[Present]) -> bool {
    let total_present_area = region
        .presents
        .iter()
        .zip(present_list.iter())
        .map(|(count, present)| count * present.cells())
        .sum();

    (region.width * region.height) >= total_present_area
}

pub fn fit_presents(input: &str) -> usize {
    let (presents, regions) = parse_input(input);
    regions.iter().filter(|region| region_fits(region, &presents)).count()
}

pub const fn have_fun(_input: &str) -> &'static str {
    "\u{2605}"
}

pub fn main() {
    let input = std::fs::read_to_string("input/day12.txt").expect("Input file not found");
    let now = std::time::Instant::now();
    println!("PART 1 = {}", fit_presents(&input));
    println!("PART 2 = {}", have_fun(&input));
    println!("Execution time: {:?}", now.elapsed());
}
